index.d.ts

```diff
@@ -0,0 +1,144 @@
+export interface Options {
+	/**
+	 * Number of concurrent pending promises. Minimum: `1`.
+	 *
+	 * @default Infinity
+	 */
+	concurrency?: number;
+}
+
+export type PromiseFactory<T> = () => PromiseLike<T>;
+
+/**
+ * Run promise-returning & async functions concurrently with optional limited concurrency.
+ *
+ * @param tasks - Iterable with promise-returning/async functions.
+ * @returns A `Promise` that is fulfilled when all promises returned from calling the functions in
+ * `tasks` are fulfilled, or rejects if any of the promises reject. The fulfilled value is an `Array`
+ * of the fulfilled values in `tasks` order.
+ */
+declare function pAll<
+	Result1,
+	Result2,
+	Result3,
+	Result4,
+	Result5,
+	Result6,
+	Result7,
+	Result8,
+	Result9,
+	Result10
```

Comment:
What if there are more than 10 inputs? There must be a better way of achieving this...

Reply:
There is a not-so-convenient fallback for any number of parameters:
```ts
declare function pAll<TAll>(
	tasks: Iterable<PromiseFactory<TAll>> | PromiseFactory<TAll>[],
	options?: Options
): Promise<TAll[]>;
```

ATM, TypeScript doesn't have a better way to describe this one.

Reply:
Do you know if there's any TS issue for this? Seems like a very big limitation.

Reply:
Seems like you've already found it, but for anyone else that's reading this there's a related issue in https://github.com/microsoft/TypeScript/issues/5453, which was addressed via variadic tuple types, but unfortunately [doesn't solve this yet](https://github.com/microsoft/TypeScript/issues/39788)